flight/Modules/VtolPathFollower/vtol_follower_control.c

```diff
@@ -72,123 +197,184 @@ int32_t vtol_follower_control_path(const float dT, const PathDesiredData *pathDe
+			guidanceSettings.EndpointDeadbandWidth,
+			guidanceSettings.EndpointDeadbandCenterGain,
+			vtol_end_m, vtol_end_r) / horiz_error_mag;
+	} else {
+		scale_horiz_error_mag = 0;
+	}
+
+	float damped_ne[2] = { errors_ned[0] * scale_horiz_error_mag,
+			errors_ned[1] * scale_horiz_error_mag };
+
+	float commands_ned[3];
 
 	// Compute desired north command velocity from position error
-	northError = hold_pos_ned[0] - cur_pos_ned[0];
-	northCommand = pid_apply_antiwindup(&vtol_pids[NORTH_POSITION], northError,
+	commands_ned[0] = pid_apply_antiwindup(&vtol_pids[NORTH_POSITION], damped_ne[0],
 	    -guidanceSettings.HorizontalVelMax, guidanceSettings.HorizontalVelMax, dT);
 
 	// Compute desired east command velocity from position error
-	eastError = hold_pos_ned[1] - cur_pos_ned[1];
-	eastCommand = pid_apply_antiwindup(&vtol_pids[EAST_POSITION], eastError,
+	commands_ned[1] = pid_apply_antiwindup(&vtol_pids[EAST_POSITION], damped_ne[1],
 	    -guidanceSettings.HorizontalVelMax, guidanceSettings.HorizontalVelMax, dT);
 
-	// Limit the maximum velocity any direction (not north and east separately)
-	float total_vel = sqrtf(powf(northCommand,2) + powf(eastCommand,2));
-	float scale = 1;
-	if(total_vel > guidanceSettings.HorizontalVelMax)
-		scale = guidanceSettings.HorizontalVelMax / total_vel;
+	if (!landing) {
+		// Compute desired down comand velocity from the position difference
+		commands_ned[2] = pid_apply_antiwindup(&vtol_pids[DOWN_POSITION], errors_ned[2],
+		    -guidanceSettings.VerticalVelMax, guidanceSettings.VerticalVelMax, dT);
+	} else {
+		// Just use the landing rate.
+		commands_ned[2] = guidanceSettings.LandingRate;
+	}
+	
+	// Limit the maximum horizontal velocity any direction (not north and east separately)
+	vtol_limit_velocity(commands_ned, guidanceSettings.HorizontalVelMax);
 
-	velocityDesired.North = northCommand * scale;
-	velocityDesired.East = eastCommand * scale;
+	velocityDesired.North = commands_ned[0];
+	velocityDesired.East = commands_ned[1];
+	velocityDesired.Down = commands_ned[2];
 
-	// Compute the desired velocity from the position difference
-	downError = hold_pos_ned[2] - cur_pos_ned[2];
-	velocityDesired.Down = pid_apply_antiwindup(&vtol_pids[DOWN_POSITION], downError,
-	    -guidanceSettings.VerticalVelMax, guidanceSettings.VerticalVelMax, dT);
-	
 	VelocityDesiredSet(&velocityDesired);	
 
-	// Indicate whether we are in radius of this endpoint
-	uint8_t path_status = PATHSTATUS_STATUS_INPROGRESS;
-	float distance2 = powf(northError, 2) + powf(eastError, 2);
-	if (distance2 < (guidanceSettings.EndpointRadius * guidanceSettings.EndpointRadius)) {
-		path_status = PATHSTATUS_STATUS_COMPLETED;
+	if (update_status) {
+		uint8_t path_status = PATHSTATUS_STATUS_INPROGRESS;
+
+		if (!landing) {
+			// Indicate whether we are in radius of this endpoint
+			// And at/above the altitude requested
+			if ((vtol_magnitude(errors_ned, 2) < guidanceSettings.EndpointRadius) &&
```

Comment:
again this needs to be both optional and settable. It also should be disabled when the user has throttle control set to false, or most users will simply never complete their path. that means it probably should be moved to a method like “vtol_control_altitude_criterion” since duplicated with line 243